### PR TITLE
[REVIEW] Add GPU-accelerated ORC Reader

### LINKED ISSUE #1222: [FEA] GPU accelerated Parquet reader
Add GPU accelerated Parquet reader in cuIO.

### LINKED ISSUE #1350: [FEA] GPU-accelerated ORC reader
**Is your feature request related to a problem? Please describe.**
Current `read_orc` in libcudf uses the CPU-based Pyarrow backend. There should be a GPU-optimized backend like for `read_parquet` that can accelerate data loads.

**Describe the solution you'd like**
Implement a GPU-accelerated ORC reader and expose under `read_orc` in the same fashion as in `read_parquet`. Re-use as much as libcudf's Parquet reader implementation as possible.

**Describe alternatives you've considered**
The non-ideal alternative is porting the Apache ORC C++ implementation to CUDA; however that has the following cons: add many files where some aren't applicable for our use case, overall architecture is CPU-oriented and not GPU-oriented, dependencies on Protobufs & Apache, and differing code style.

**Additional context**
See #1222 for libcudf Parquet implementation.

### PR DESCRIPTION
**Pull request by j-ieong**
Fixes #1350.
This adds an interface for using GPU-accelerated reading and converting of Apache ORC to cuDF.

List of changes:
- adds a simple GPU-accelerated ORC-to-cuDF reader
- adds a top-level read_parquet interface and associated python bindings
- adds an engine parameter to select between pyarrow and cudf implementations
- adds to the existing parameterized pytest to check against pyarrow reference

### CODE REVIEW

File: cpp/src/io/orc/orc_reader.cu [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+                                       args->use_cols + args->use_cols_len);
+    int index = 0;
+    for (const auto &use_name : use_names) {
+      for (int i = 0; i < md.get_num_columns(); ++i, ++index) {
+        if (index >= md.get_num_columns()) {
+          index = 0;
+        }
+        if (md.ff.GetColumnName(index) == use_name) {
+          selected_cols.emplace_back(index);
+          index++;
```
Comment (OlivierNV) [inline cpp/src/io/orc/orc_reader.cu]: Missing break ?

Comment (j-ieong) [reply cpp/src/io/orc/orc_reader.cu]: Hrmm true... I guess we can't get duplicate column names in ORC?

Comment (OlivierNV) [reply cpp/src/io/orc/orc_reader.cu]: I'd imagine we'd want to still pick only the first matching one if there were (I was mainly concerned about perf here, like in the GFN dataset with 7000+ columns)

**Comment by randerzander** [comment]
@OlivierNV I see that you're decoding decimal types as int64s. Why ints instead of float64?

**Comment by OlivierNV** [comment]
> @OlivierNV I see that you're decoding decimal types as int64s. Why ints instead of float64?

Mainly to match with Apache ORC. It's either int64 or int128 (which we don't yet have in GDF), I was also initially thinking that float64 would make more sense, but the decimal type is apparently mainly used for large integers.

**Comment by randerzander** [comment]
> > @OlivierNV I see that you're decoding decimal types as int64s. Why ints instead of float64?
> 
> Mainly to match with Apache ORC. It's either int64 or int128 (which we don't yet have in GDF), I was also initially thinking that float64 would make more sense, but the decimal type is apparently mainly used for large integers.

We should consider exposing an option to convert to float64 intstead, as there are users who actually encode decimal data instead of big integers.

File: cpp/src/io/orc/orc_reader.cu (hunk 1 above)
Comment (j-ieong) [reply cpp/src/io/orc/orc_reader.cu]: Okay, fixed.

**Comment by OlivierNV** [comment]
> We should consider exposing an option to convert to float64 intstead, as there are users who actually encode decimal data instead of big integers.

Yes, in fact, I think the initial version will just map decimals to float64, since cudf does not yet have a global column scale or int128 support (should be ready in the next few days).

**Comment by OlivierNV** [comment]
rerun tests

File: python/cudf/io/orc.py [hunk 2]
```diff
@@ -1,17 +1,30 @@
...
 from cudf.utils import ioutils
 
 import pyarrow.orc as orc
+
 import warnings
 
 
 @ioutils.doc_read_orc()
-def read_orc(path, columns=None, **kwargs):
+def read_orc(path, engine='pyarrow', columns=None, skip_rows=None,
```
Comment (randerzander) [inline python/cudf/io/orc.py]: Shouldn't we change the default engine to 'orc' now?

Comment (j-ieong) [reply python/cudf/io/orc.py]: Yup, fixed. Thanks!

**Comment by harrism** [comment]
@j-ieong Since this is a big PR and we are in burn down, I'm happy to defer my review to @mjsamoht since he is cuIO PIC and your previous PRs have all been top-notch.

If there are specific things you'd like me to look at please let me know.

**Pull request merged**